meadow-client/src/Halogen/Blockly.purs

```diff
@@ -0,0 +1,130 @@
+import Data.Symbol (SProxy(..))
+import Effect.Class (class MonadEffect)
+import Effect.Class.Console (log)
+import Halogen (ClassName(..), Component, ComponentDSL, ComponentHTML, RefLabel(..), action, get, lifecycleComponent, liftEffect, modify, raise)
+import Halogen.HTML (button, div, text)
+import Halogen.HTML as HH
+import Halogen.HTML.Events (input_, onClick)
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties (classes, id_, ref)
+import Marlowe.Blockly (buildBlocks, buildGenerator)
+import Marlowe.Parser as Parser
+import Marlowe.Pretty (pretty)
+import Marlowe.Types (Contract(..))
+import Prelude (Unit, bind, discard, pure, show, unit, ($))
+import Text.Parsing.Parser (runParser)
+import Text.Parsing.Parser.Basic (parens)
+
+type BlocklyState
+  = {blocklyState :: Maybe BT.BlocklyState, generator :: Maybe Generator}
+
+_blocklyState :: Lens' BlocklyState (Maybe BT.BlocklyState)
+_blocklyState = prop (SProxy :: SProxy "blocklyState")
+
+_generator :: Lens' BlocklyState (Maybe Generator)
+_generator = prop (SProxy :: SProxy "generator")
+
+data BlocklyQuery a
+  = Inject (Array BlockDefinition) a
+  | Resize a
+  | SetData Unit a
+  | GetCode a
+  | SetCode String a
+
+data BlocklyMessage
+  = Initialized
+  | CurrentCode String
+
+type HTML
+  = ComponentHTML BlocklyQuery
+
+type DSL m
+  = ComponentDSL BlocklyState BlocklyQuery BlocklyMessage m
+
+blockly :: forall m. MonadEffect m => Array BlockDefinition -> Component HH.HTML BlocklyQuery Unit BlocklyMessage m
+blockly blockDefinitions =
+  lifecycleComponent
+    { initialState: \_ -> {blocklyState: Nothing, generator: Nothing}
+    , render
+    , eval
+    , initializer: Just $ action (Inject blockDefinitions)
+    , finalizer: Nothing
+    , receiver: HE.input SetData
+    }
+
+eval :: forall m. MonadEffect m => BlocklyQuery ~> DSL m
+eval (Inject blockDefinitions next) = do
+  blocklyState <- liftEffect $ Blockly.createBlocklyInstance (ElementId "blocklyWorkspace") (ElementId "blocklyToolbox")
+  liftEffect $ Blockly.addBlockTypes blocklyState blockDefinitions
+  let _ = ST.run (do
+                    workspaceRef <- STRef.new blocklyState.workspace
+                    Blockly.initializeWorkspace blocklyState.blockly workspaceRef)
+  let generator = buildGenerator blocklyState
+  _ <- modify _ {blocklyState = Just blocklyState, generator = Just generator}
+  pure next
+
+eval (Resize next) = do
+  state <- get
+  _ <- liftEffect $ Blockly.resize state.blocklyState
+  pure next
+
+eval (SetData _ next) = pure next
+
+eval (GetCode next) = runMaybeT f *> pure next
+  where
+  f = do
+    blocklyState <- MaybeT $ use _blocklyState
+    generator <- MaybeT $ use _generator
+    let code = workspaceToCode blocklyState generator
+        result = runParser code (parens Parser.contract <|> Parser.contract)
+    lift $ case result of
+        Left e -> log (show e)
+        Right contract -> raise $ CurrentCode (show (pretty contract))
```

Comment:
Probably we should show an error when this fails (other than in the JS console). I think it should only fail when there is gaps in the blockly puzzle.
The original Meadow didn't show anything either, but people did complain about it.

Reply:
Ah, so this is because it should _never_ fail this part. If there are gaps then clicking the button does nothing at the moment.

Reply:
There is another type of failure mode when there are empty blocks or unconnected blocks. In these cases we currently log to the console and do nothing but that is defined in other code. Should we show an error message?

Reply:
ok, now all expected errors with blockly2code are reported to the user as "This workspace cannot be converted to code" and unexpected errors are reported to the user with ""An unexpected error has occurred, please raise a support issue: (whatever the raw error is)"

Reply:
Nice :)

For unconnected blocks we should ignore them if possible. Currently I tried with an unconnected Null block and it takes that as the main contract.